lib/dentaku/ast/functions/ifs.rb

```diff
@@ -0,0 +1,64 @@
+require_relative '../function'
+
+module Dentaku
+  module AST
+    class Ifs < Function
+      # This function (based off excel IFS function) was requested by Matt Cloyd 
+      # https://support.microsoft.com/en-us/office/ifs-function-36329a26-37b2-467c-972b-4a39bd951d45
+      #
+      # Example
+      # ifs(
+      #   a > 1, 100,
+      #   b > 100, 200,
+      #   b < 100, 300,
+      #   true, 0 # default case
+      # )
+      def self.type_syntax
+        # We want something like
+        # "ifs(*[:bool, %a]...) = %a"
+        # but not sure if it's even reasonable to get our type system to handle something like an unending sequence of types (:bool, %a)
+        #
+        # we have other places where a tuple would be handy. a list that can support multiple types, but it's length and types are fixed
```

Comment:
There have been a few cases this has come up, but I'm not remembering right now.

Would we need something along these lines for the `case` `in` pattern?
e.g.
```
case [field:use_category, field:construction_type]
in [use_category:residential, enum:new] then 1000
in [use_category:business, enum:new] then 1500
in [use_category:business, enum:remodel] then 2500
in [use_category:business] then 5000
else 10000
end
```

Reply:
This should be perfectly possible in the type system `ifs([[:bool, %a]], %a) = %a`

Reply:
oh wait nvm we don't have tuples

Reply:
our dictionaries have separate types though, so `ifs([{cond: :bool, body: %a}], %a) = %a` would work, buuuuut idk

Reply:
We do have quite some flexibility on the `:param` type though - it's any parameterized type. This works just fine:
``` ruby
module Dentaku
  class Pair
    attr_reader :l, :r
    def initialize(l, r) @l = l; @r = r end
    def inspect
      "pair(#{l.inspect}, #{r.inspect})"
    end
  end
end

Dentaku::AST::Function.register('pair(%a, %b) = :pair(%a %b)', ->(a, b) { Dentaku::Pair.new(a, b) })
Dentaku::AST::Function.register('left(:pair(%a, %b)) = %a', ->(t) { t.l })
Dentaku::AST::Function.register('right(:pair(%a, %b)) = %b', ->(t) { t.r })
```

```
[8] pry(main)> Dentaku::Type::Checker.new { }.check!(Dentaku::Calculator.new.ast('pair(1, "a")'))
=> <SolutionSet
<Constraint ["a"] = :string>
<Constraint %b14 = :string>
<Constraint [1] = :numeric>
<Constraint %a13 = :numeric>
<Constraint [pair(1, "a")] = :pair(:numeric :string)> >
[9] pry(main)> Dentaku::Calculator.new.evaluate('pair(1, "a")')
=> pair(1, "a")
[10] pry(main)> Dentaku::Calculator.new.evaluate('left(pair(1, "a"))')
=> 1
[11] pry(main)> Dentaku::Calculator.new.evaluate('right(pair(1, "a"))')
=> "a"
```

Reply:
Also! I think this would do the same thing:

```
CASE true
WHEN condition1 THEN result1
WHEN condition2 THEN result2
ELSE default
END
```